File: python/src/toolrate_mcp/tools/fallback_chain.py
"""``toolrate_fallback_chain`` — discover what agents switch to when a tool fails."""
from __future__ import annotations

import json
from typing import Annotated, Any, Optional

from pydantic import Field

from ..client import api_request
from ..errors import raise_tool_error

NAME = "toolrate_fallback_chain"

DESCRIPTION = "\n".join(
    [
        "Given a tool identifier, return the tools that real agents actually switched to after this one failed — ranked by success rate in the fallback position.",
        "",
        "Use this when toolrate_assess flagged the primary tool as risky, or when a live call just failed and you need a fallback.",
    ]
)


async def handler(
    tool_identifier: Annotated[str, Field(min_length=1, max_length=512, description="The tool to find fallbacks for.")],
    limit: Annotated[Optional[int], Field(default=None, ge=1, le=20, description="Max results (default 5).")] = None,
) -> str:
    try:
        body, rate_limit = await api_request(
            method="GET",
            path="/v1/discover/fallback-chain",
            query={"tool_identifier": tool_identifier, "limit": limit},
        )
    except Exception as e:
        raise_tool_error(e)

    result: dict[str, Any] = {**body}
    if rate_limit is not None:
        result["_meta"] = {"rate_limit": rate_limit}
    return json.dumps(result, indent=2)

File: typescript/src/tools/fallback_chain.ts
import { apiRequest } from "../client.js";
import { mcpError } from "../errors.js";
import type { RawFallbackChainResponse } from "../schema.js";
import { fallbackChainInputSchema, type FallbackChainInput } from "../types.js";

export const name = "toolrate_fallback_chain";

export const config = {
  title: "Find the best alternatives when a specific tool fails",
  description: [
    "Given a tool identifier, return the tools that real agents actually switched to after this one failed — ranked by success rate in the fallback position.",
    "",
    "Use this when toolrate_assess flagged the primary tool as risky, or when a live call just failed and you need a fallback.",
  ].join("\n"),
  inputSchema: fallbackChainInputSchema,
};

export async function handler(input: FallbackChainInput) {
  try {
    const { body, rateLimit } = await apiRequest<RawFallbackChainResponse>({
      method: "GET",
      path: "/v1/discover/fallback-chain",
      query: {
        tool_identifier: input.tool_identifier,
        limit: input.limit,
      },
    });

    const result = {
      ...body,
      ...(rateLimit ? { _meta: { rate_limit: rateLimit } } : {}),
    };

    return {
      content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }],
    };
  } catch (err) {
    return mcpError(err);
  }
}
